pkg/privilege/privileges/cache.go

```diff
@@ -953,229 +974,269 @@ func parseHostIPNet(s string) *net.IPNet {
 func (record *baseRecord) assignUserOrHost(row chunk.Row, i int, f *resolve.ResultField) {
 	switch f.ColumnAsName.L {
 	case "user":
-		record.User = row.GetString(i)
+		record.User = strings.Clone(row.GetString(i))
 	case "host":
-		record.Host = row.GetString(i)
+		record.Host = strings.Clone(row.GetString(i))
 		record.patChars, record.patTypes = stringutil.CompilePatternBinary(record.Host, '\\')
 		record.hostIPNet = parseHostIPNet(record.Host)
 	}
 }
 
-func (p *MySQLPrivilege) decodeUserTableRow(row chunk.Row, fs []*resolve.ResultField) error {
-	var value UserRecord
-	defaultAuthPlugin := ""
-	if p.globalVars != nil {
-		val, err := p.globalVars.GetGlobalSysVar(vardef.DefaultAuthPlugin)
-		if err == nil {
-			defaultAuthPlugin = val
-		}
-	}
-	if defaultAuthPlugin == "" {
-		defaultAuthPlugin = mysql.AuthNativePassword
-	}
-	for i, f := range fs {
-		switch {
-		case f.ColumnAsName.L == "authentication_string":
-			value.AuthenticationString = row.GetString(i)
-		case f.ColumnAsName.L == "account_locked":
-			if row.GetEnum(i).String() == "Y" {
-				value.AccountLocked = true
-			}
-		case f.ColumnAsName.L == "plugin":
-			if row.GetString(i) != "" {
-				value.AuthPlugin = row.GetString(i)
-			} else {
-				value.AuthPlugin = defaultAuthPlugin
-			}
-		case f.ColumnAsName.L == "token_issuer":
-			value.AuthTokenIssuer = row.GetString(i)
-		case f.ColumnAsName.L == "user_attributes":
-			if row.IsNull(i) {
-				continue
-			}
-			bj := row.GetJSON(i)
-			pathExpr, err := types.ParseJSONPathExpr("$.metadata.email")
-			if err != nil {
-				return err
+func (p *MySQLPrivilege) decodeUserTableRow(userList map[string]struct{}) func(chunk.Row, []*resolve.ResultField) error {
+	return func(row chunk.Row, fs []*resolve.ResultField) error {
```

Comment:
The decoder function is changed from 

```
func (p *MySQLPrivilege) decodeUserTableRow(row chunk.Row, fs []*resolve.ResultField) error
```

to 

```
func (p *MySQLPrivilege) decodeUserTableRow(userList map[string]struct{}) func(chunk.Row, []*resolve.ResultField) error
```

Because now we need the user = 'xxx' filter condition during decoding...
The code seems to be changed a lot but it's really just function signature changes.